SimTracker/TrackerHitAssociation/plugins/SimDoubletsProducer.cc

```diff
@@ -0,0 +1,271 @@
+  edm::ParameterSetDescription descTPSelector;
+  descTPSelector.add<double>("ptMin", 0.9);
+  descTPSelector.add<double>("ptMax", 1e100);
+  descTPSelector.add<double>("minRapidity", -3.);
+  descTPSelector.add<double>("maxRapidity", 3.);
+  descTPSelector.add<double>("tip", 60.);
+  descTPSelector.add<double>("lip", 30.);
+  descTPSelector.add<int>("minHit", 0);
+  descTPSelector.add<bool>("signalOnly", true);
+  descTPSelector.add<bool>("intimeOnly", false);
+  descTPSelector.add<bool>("chargedOnly", true);
+  descTPSelector.add<bool>("stableOnly", false);
+  descTPSelector.add<std::vector<int>>("pdgId", {});
+  descTPSelector.add<bool>("invertRapidityCut", false);
+  descTPSelector.add<double>("minPhi", -3.2);
+  descTPSelector.add<double>("maxPhi", 3.2);
+  desc.add<edm::ParameterSetDescription>("TrackingParticleSelectionConfig", descTPSelector);
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+// Phase 2 fillDescription
+template <>
+void SimDoubletsProducer<pixelTopology::Phase2>::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+
+  // sources for cluster-TrackingParticle association, TrackingParticles, RecHits and the beamspot
+  desc.add<edm::InputTag>("clusterTPAssociationSrc", edm::InputTag("hltTPClusterProducer"));
+  desc.add<edm::InputTag>("trackingParticleSrc", edm::InputTag("mix", "MergedTrackTruth"));
+  desc.add<edm::InputTag>("pixelRecHitSrc", edm::InputTag("hltSiPixelRecHits"));
+  desc.add<edm::InputTag>("beamSpotSrc", edm::InputTag("hltOnlineBeamSpot"));
+
+  // parameter set for the selection of TrackingParticles that will be used for SimHitDoublets
+  edm::ParameterSetDescription descTPSelector;
+  descTPSelector.add<double>("ptMin", 0.9);
+  descTPSelector.add<double>("ptMax", 1e100);
+  descTPSelector.add<double>("minRapidity", -4.5);
+  descTPSelector.add<double>("maxRapidity", 4.5);
+  descTPSelector.add<double>("tip", 2.);  // NOTE: differs from HLT MultiTrackValidator
+  descTPSelector.add<double>("lip", 30.);
+  descTPSelector.add<int>("minHit", 0);
+  descTPSelector.add<bool>("signalOnly", true);
+  descTPSelector.add<bool>("intimeOnly", false);
+  descTPSelector.add<bool>("chargedOnly", true);
+  descTPSelector.add<bool>("stableOnly", false);
+  descTPSelector.add<std::vector<int>>("pdgId", {});
+  descTPSelector.add<bool>("invertRapidityCut", false);
+  descTPSelector.add<double>("minPhi", -3.2);
+  descTPSelector.add<double>("maxPhi", 3.2);
+  desc.add<edm::ParameterSetDescription>("TrackingParticleSelectionConfig", descTPSelector);
+
+  descriptions.addWithDefaultLabel(desc);
+}
+
+template <typename TrackerTraits>
+void SimDoubletsProducer<TrackerTraits>::beginRun(const edm::Run& run, const edm::EventSetup& eventSetup) {
+  // get TrackerGeometry and TrackerTopology
+  trackerGeometry_ = &eventSetup.getData(geometry_getToken_);
+  trackerTopology_ = &eventSetup.getData(topology_getToken_);
```

Comment:
Only getting EventSetup data products in (e.g.) beginRun is an anti-pattern. The `EventSetup::getData()` is fast, and is recommended to be done in the `produce()` function.